drivers/wifi/esp8266/wifi_esp8266.c

```diff
@@ -0,0 +1,1570 @@
+		}
+
+		printk(" [%s]\n", output);
+	}
+}
+
+
+/* Send an AT command with a series of response handlers */
+static int send_at_cmd(struct socket_data *sock,
+			const u8_t *data, int timeout)
+{
+	int ret;
+
+	foo_data.last_error = 0;
+
+	k_mutex_lock(&dev_mutex, K_FOREVER);
+	k_sem_reset(&foo_data.response_sem);
+	mdm_receiver_send(&foo_data.mdm_ctx, data, strlen(data));
+	mdm_receiver_send(&foo_data.mdm_ctx, "\r\n", 2);
+	k_mutex_unlock(&dev_mutex);
+
+	if (timeout == K_NO_WAIT) {
+		return 0;
+	}
+
+	ret = k_sem_take(&foo_data.response_sem, timeout);
+
+	if (ret == 0) {
+		ret = foo_data.last_error;
+	} else if (ret == -EAGAIN) {
+		ret = -ETIMEDOUT;
+	}
+
+	return ret;
+}
+
+static int net_buf_find_next_delimiter(struct net_buf *buf, const u8_t d,
+	size_t index)
+{
+	struct net_buf *frag = buf;
+	u16_t offset = 0;
+	u16_t pos = 0;
+
+	while (frag && index) {
+		if (frag->len > index) {
+			offset += index;
+			pos += index;
+			break;
+		} else {
+			index -= frag->len;
+			pos += frag->len;
+			offset = 0;
+			if (!frag->frags) {
+				return -1;
+			}
+			frag = frag->frags;
+		}
+	}
+
+	while(*(frag->data + offset) != d) {
+		if (offset == frag->len) {
+			if (!frag->frags) {
+				return -1;
+			}
+			frag = frag->frags;
+			offset = 0;
+		} else {
+			offset++;
+		}
+		pos++;
+	}
+
+	return pos;
+}
+
+static int net_buf_ncmp(struct net_buf *buf, const u8_t *s2, size_t n)
+{
+	struct net_buf *frag = buf;
+	u16_t offset = 0;
+
+	while ((n > 0) && (*(frag->data + offset) == *s2) && (*s2 != '\0')) {
+		if (offset == frag->len) {
+			if (!frag->frags) {
+				break;
+			}
+			frag = frag->frags;
+			offset = 0;
+		} else {
+			offset++;
+		}
+
+		s2++;
+		n--;
+	}
+
+	return (n == 0) ? 0 : (*(frag->data + offset) - *s2);
+}
+
+/* Echo Handler for commands without related sockets */
+static void on_cmd_atcmdecho_nosock(struct net_buf **buf, u16_t len)
+{
+}
+
+static void on_cmd_esp8266_ready(struct net_buf **buf, u16_t len)
+{
+	k_sem_give(&foo_data.response_sem);
+}
+
+static void esp8266_ip_addr_work(struct k_work *work)
+{
+	int ret;
+
+	ret = send_at_cmd(NULL, "AT+CIPSTA_CUR?", MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("failed to get ip address information\n");
+		return;
+	}
+
+	/* update interface addresses */
+	net_if_ipv4_set_gw(foo_data.iface, &foo_data.gw);
+	net_if_ipv4_set_netmask(foo_data.iface,
+			&foo_data.netmask);
+	net_if_ipv4_addr_add(foo_data.iface, &foo_data.ip, NET_ADDR_DHCP,
+		0);
+}
+
+static void on_cmd_ip_addr_get(struct net_buf **buf, u16_t len)
+{
+		k_delayed_work_submit_to_queue(&esp8266_workq,
+			&foo_data.ip_addr_work, K_SECONDS(2));
+}
+
+static void on_cmd_conn_open(struct net_buf **buf, u16_t len)
+{
+	int id;
+	u8_t msg[2];
+
+	net_buf_linearize(msg, 1, *buf, 0, 1);
+	msg[1] = '\0';
+	id = atoi(msg);
+
+	foo_data.socket_data[id].connected = 1;
+}
+
+static void on_cmd_conn_closed(struct net_buf **buf, u16_t len)
+{
+	int id;
+	u8_t msg[2];
+	struct socket_data *sock = NULL;
```

Comment:
usually, in net side, we put pre-assigned variables first in the list of declaration.